company-lsp.el

```diff
@@ -511,6 +543,56 @@ CALLBACK is a function that takes a list of strings as completion candidates."
                                      (funcall callback (company-lsp--on-completion resp prefix)))))
     (setq company-lsp--last-request-id (plist-get body :id))))
 
+(defun company-lsp--make-completion-request (prefix)
+  "Make request body for completion.
+
+PREFIX is a string prefix given by company-mode.
+
+Returns the request body that can be used by `lsp-send-request'
+or `lsp-send-request-async'."
+  (let ((params (lsp--text-document-position-params)))
+    (when company-lsp-enable-trigger-kind
+      (setq params (plist-put params :context
+                              `(:triggerKind ,(company-lsp--get-trigger-kind prefix)))))
+    (lsp--make-request "textDocument/completion"
+                       params)))
+
+(defun company-lsp--get-trigger-kind (prefix)
+  "Determine why completion is triggered for PREFIX.
+
+Returns one of `company-lsp--trigger-kind-invoked',
+`company-lsp--trigger-kind-trigger-character' and
+`company-lsp--trigger-kind-incomplete'."
+  (cond
+   ((or (eq this-command 'company-lsp)
+        (eq this-command 'company-begin-backend))
+    ;; Explicitly calling completion command.
+    company-lsp--trigger-kind-invoked)
+   ((company-lsp--cache-find-closest prefix)
+    ;; Has incomplete candidates for sub-prefix. This assumes that if the
+    ;; candidates for sub-prefix, completion won't reach this step since the
+    ;; candidates for the sub-prefix can be narrowed down and returned directly.
+    company-lsp--trigger-kind-incomplete)
+   ((or (null company-point) (< company-point (point)))
```

Comment:
This is probably an okay approach, at least for the near future.

Eglot's solution seems more future-proof, though: https://github.com/joaotavora/eglot/blob/7d6e3cf5d7ae098aa6c8572343c8bc9b8453aace/eglot.el#L1504-L1510

Reply:
The reason why I prefer this one is that there is no way for company-lsp to remove its hooks. There is no `lsp-mode-disable-hook` or `company-mode-disable-hook` for it and `company-lsp` itself is not a minor hook.